### PR TITLE
coercion: use a generalization of the target to specialize the source to coerce instead of propagating the wrong type.

### PR DESCRIPTION
**Pull request by eddyb**
The issue solved here involved generic "constructor" traits and unsizing coercions.
A real usecase is `box` desugaring, which is being demonstrated as a test-case.

Given such a trait with more than one implementation, e.g. for `Box<T>` and `Rc<T>`:

``` rust
trait Make<T> {
    fn make(T) -> Self;
}
impl<T> Make<T> for Box<T> {
    fn make(x: T) -> Box<T> { Box::new(x) }
}
impl<T> Make<T> for Rc<T> {
    fn make(x: T) -> Rc<T> { Rc::new(x) }
}
```

An direct call would be too unconstrained for a coercion, so the expected type ends up propagated:

``` rust
fn make_boxed_trait<E: Error>(x: E) -> Box<Error> {
    // error: Make<E> not implemented for Box<Error>.
    Make::make(x)
}
```

The workaround was to manually specify enough about the produced type to allow resolving it:

``` rust
fn make_boxed_trait<E: Error>(x: T) -> Box<Error> {
    let boxed: Box<_> = Make::make(x);
    // <Box<_> as Make<E>> gets selected to the appropriate impl, and so
    // boxed has the type Box<E> at this point, which can coerce to Box<Error>.
   boxed
}
```

That workaround is implemented in this PR, for all coercions where unsizing was ambiguous.
A generalized type is picked based on the `CoerceUnsize` impl that could be used for coercing.

For example, `CoerceUnsized<Box<Error>>` is implemented by `Box<T> forall T`, so the generalized
type is `Box<_>`, which matches both `Box<E>` and `Box<Error>`.
Such generalization allows constraining the source of the potential coercion while still allowing
it to be any type that could coerce to the target, or the target type itself.

### CODE REVIEW

**Comment by pnkfelix** [comment]
@eddyb I'm confused by the description you have in this issue.

When you write "That workaround is implemented in this PR", I now infer that you mean: "This PR makes explicitly writing such workarounds in one's code unnecessary, because the compiler will now effectively do the work for you."

So, before I jump into the review itself, is that a correct inference ?

(And if so, maybe rephrase the description so that the commit of the code to the repo will have a somewhat clearer statement of its purpose?)

**Comment by pnkfelix** [comment]
@eddyb oh, no, wait, my inference above was incorrect, I guess?  That is, you *do* still need to employ the described workaround in the user's code.

**Comment by eddyb** [comment]
@pnkfelix I did expect to need changes to the explanations, as the feature itself is a bit convoluted.
Your first interpretation is correct, let me try to come up with a better phrasing.
How about "This PR enables the compiler to generate the same coercion, without the above workaround being employed in the code"?

**Comment by pnkfelix** [comment]
@eddyb yes, that sounds better.  I still need to read the test case more carefully, I'll have more feedback shortly.

**Comment by pnkfelix** [comment]
It may be even nicer to have a test that shows the kind of case that we *really* expect to come up, namely where no type annotation is present at all on any of the local variables, and instead the constraints like `Box<_>` or `Rc<_>` on the type arise from where the value flows to, e.g. return position for the demo function, or argument position for some subroutine that the demo function invokes.

Having said that, at least I grok the test at this point.

File: src/librustc_typeck/check/coercion.rs [hunk 1]
```diff
@@ -121,39 +131,107 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
...
+                return self.coerce_from_fn_item(a, a_f, b);
+            }
+            ty::TyBareFn(None, a_f) => {
+                // We permit coercion of fn pointers to drop the
+                // unsafe qualifier.
+                return self.coerce_from_fn_pointer(a, a_f, b);
+            }
+            _ => {}
+        }
+
```
Comment (pnkfelix) [inline src/librustc_typeck/check/coercion.rs]: so, in the code before this PR, the control flow that reached this point (line 159) corresponds to the comment above "Otherwise, just use subtyping rules."

Clearly you are no longer "just using the subtyping rules" in the code that follows that point in the control flow here.  :)

Still, for some reason I wish there was a comment line or two here of the same pithy nature as "Otherwise just use subtyping rules" before jumping straight into the discussion of "generalization in hopes of an unsizing coercion" ... something that helped the reader re-establish what case we are in in the coercion analysis?  Or is that just a pipe dream?

File: src/librustc_typeck/check/coercion.rs [hunk 2]
```diff
@@ -60,28 +60,44 @@
...
 
-use std::cell::RefCell;
 use std::collections::VecDeque;
 use syntax::ast;
 
 struct Coerce<'a, 'tcx: 'a> {
     fcx: &'a FnCtxt<'a, 'tcx>,
-    origin: infer::TypeOrigin,
-    unsizing_obligations: RefCell<Vec<traits::PredicateObligation<'tcx>>>,
+    origin: infer::TypeOrigin
```
Comment (pnkfelix) [inline src/librustc_typeck/check/coercion.rs]: super nitpick: please leave the trailing comma here, if only to reduce the diff of both this PR and future ones.

File: src/librustc_typeck/check/coercion.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc_typeck/check/coercion.rs]: Something along the lines of "All attempts at coercion have failed, must use subtyping (or try harder)."?

File: src/librustc_typeck/check/coercion.rs [hunk 3]
```diff
@@ -121,39 +131,107 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
...
+            }
+            ty::TyBareFn(None, a_f) => {
+                // We permit coercion of fn pointers to drop the
+                // unsafe qualifier.
+                return self.coerce_from_fn_pointer(a, a_f, b);
+            }
+            _ => {}
+        }
+
+        // Attempt to generalize the expected type in hopes of an unsizing
```
Comment (pnkfelix) [inline src/librustc_typeck/check/coercion.rs]: It may be just my own personal style, but for a comment like this that is so embedded in the problem of dealing with the `box expr` desugaring, it may be good to actually include a link to a relevant github issue so that someone who is trying to understand this has an alternative place to look for help besides this comment.

E.g. you could link to Issue #22405 (which, incidentally, should be closed by this PR, right?)

Comment (eddyb) [inline src/librustc_typeck/check/coercion.rs]: Thanks! I completely forgot about that specific issue (and did not consider searching for one for some reason).

File: src/librustc_typeck/check/coercion.rs [hunk 4]
```diff
@@ -121,39 +131,107 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
...
+        // does not implement `Make`, but `Box<T>` does. The stop-gap fix
+        // is `Make::make(x: T): Box<T>: Box<Trait>`.
+        // In that same case, the following generalization will attempt to
+        // apply `Box<_>` to the otherwise unconstrained `Make::make` return
+        // type and trigger selection, hoping to get the unambiguous source
+        // type `Box<T>` for the coercion to `Box<Trait>`.
+        // Subtyping rules are used if the generalization and second attempt
+        // at coercions through unsizing do not apply.
+
+        // The first unsizing coercion must have failed due to ambiguity.
```
Comment (pnkfelix) [inline src/librustc_typeck/check/coercion.rs]: this comment ... I'm having a hard time reconciling it with the code.  Namely the phrase "must have failed", and the fact that the condition in the test is checking if the err is *not* `CoerceUnsizedError::Ambiguous`...

If it were "*may* have failed"  then I would understand (but would also deem the comment entirely redundant).

But as it is, there are three cases for `err()`: `None`, `Some(Inapplicability)` or `Some(Ambiguity)`.  We can't be in the `None` case because we already handled `Ok` via the `if let` earlier (right?).  But that still leaves `Inapplicability` versus `Ambiguity`.

(Is the comment just an dead artifact that I should be ignoring?)

Comment (eddyb) [inline src/librustc_typeck/check/coercion.rs]: The "must" refers to a condition for trying the generalization.
If the first unsizing coercion did not "fail due to ambiguity", then it "failed" due to certain inapplicability, so there is no point in trying harder.
The use of "fail" here was meant to be about the fact that the compiler didn't go through with the unsizing coercion at the start of the process.

Comment (pnkfelix) [inline src/librustc_typeck/check/coercion.rs]: Oh I see!  Okay, I would write this as: "We only apply the aforementioned generalization if the first attempt at an unsizing coercion failed due to ambiguity."

**Comment by eddyb** [comment]
I have found a really simple example that will be mishandled by the current implementation:
```rust
impl CoerceUnsized<Pointer<Trait>> for Pointer<Trait> {}
impl<T: Trait> CoerceUnsized<Pointer<Trait>> for Pointer<T> {}
```
Selecting `<Pointer<Trait> as CoerceUnsized<Pointer<Trait>>>` won't be ambiguous, and it would result in the first `impl.
It's harmless, because it will result in the same behavior as before this patch, but it would be better to detect this ambiguity (by manually looking at `impl`s, instead of using trait selection) and maybe even find the common supertype of the `Self` types of all applicable `impl`s (`Pointer<_>` in this case).

**Comment by pnkfelix** [comment]
I've skimmed this but I really want @nikomatsakis to look at it too, since I think has expressed concern over the growing complexity of the coercion logic.

r? @nikomatsakis

**Comment by nikomatsakis** [comment]
Sorry, I'm being slow again. I plan to examine this this week. I have to admit I am pretty wary here, though, I sort of want to tread lightly on inference until we have a feeling for the full set of use cases that we aim to support. Modifying the unsizing coercion seems like an interesting approach but not clearly the right one -- and these sorts of changes are very hard to "contain" in terms of stabilization.

**Comment by alexcrichton** [comment]
ping? needs a review?

**Comment by nikomatsakis** [comment]
Yes, I've been very overwhelmed. I'm working on clearing my backlog right now, including this PR.

File: src/librustc_typeck/check/coercion.rs [hunk 5]
```diff
@@ -60,28 +60,44 @@
...
 }
 
 type CoerceResult<'tcx> = RelateResult<'tcx, Option<ty::AutoAdjustment<'tcx>>>;
 
+/// The result of an attempt at coercing a Source type to a
+/// Target type via unsizing (`Source: CoerceUnsized<Target>`).
+/// If successful, all `CoerceUnsized` and `Unsized` obligations were selected.
+/// Other obligations, such as `T: Trait` for `&T -> &Trait`, are provided
+/// alongside the adjustment, to be enforced later.
+type CoerceUnsizedResult<'tcx> = Result<(AutoDerefRef<'tcx>,
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/coercion.rs]: Nit: this type is getting complex enough that adding a struct to represent the Ok case might be a good idea, or else using a different enum altogether.

**Comment by nikomatsakis** [comment]
/me starts reading this PR. Sorry for the delay.

File: src/librustc_typeck/check/coercion.rs (hunk 5 above)
Comment (eddyb) [inline src/librustc_typeck/check/coercion.rs]: I believe I had tried that at some point and it got unwieldy, but maybe just a `struct` would work.

**Comment by nikomatsakis** [comment]
(OK, still reading, sorry, this is dense code I'm having to bring it back into cache)

**Comment by eddyb** [comment]
@nikomatsakis Any progress on this? I almost feel like we should move to https://reviewable.io/ so we can keep track of reviews.

**Comment by nikomatsakis** [comment]
A meta-comment: I think that changes like this ought to be feature-gated in some way, like any other. This one seems particularly amenable to such treatment.

**Comment by nikomatsakis** [comment]
@eddyb not sure if you saw my comments on this; sorry it took me so amazingly long to read it. The main reason for the delay is that I feel like we need to get a better handle on coercions before we go and expand them, it's hard for me to tell if this approach is good or is something that will box us into a corner. A feature gate would make me feel a lot better about that. All that said, there is only so long we need to stall here-- I admit this patch is a lot cleaner and more elegant than I initially anticipated, and this may well be a part of the way forward.

Perhaps a good idea would be to schedule a time to brainstorm and chat about this as a video chat. I'd also like to sketch out how this affects other scenarios that seem strongly related; I have a few in mind but have to go scouring logs to find the specific examples.

**Comment by alexcrichton** [comment]
(ping)

Should this be rebased? Closed? 

(just clearing out the queue)

**Comment by nikomatsakis** [comment]
@alexcrichton Heh. This is an old PR by now. @eddyb and I still have not caught up as I described in my comment -- but maybe we can make time for it. I've forgotten the details of this PR now but I remember it was nicer than I feared. :) 

This part of the codebase hasn't really changed much in the meantime though, I imagine rebasing would not be so hard.

**Comment by eddyb** [comment]
@nikomatsakis Actually, I've abandoned this approach as there are other issues, and [there is a much more principled approach](https://github.com/rust-lang/rfcs/pull/1426#issuecomment-167010445) to the whole thing.
In retrospect, I should've mentioned you in that precise comment.

**Comment by nikomatsakis** [comment]
No worries, I'll take a look now:)
